designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,224 @@
+# Disruption Controls By Action 
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. High-Frequency Trading (HFT) firms require full compute capacity during specific operational hours, making it imperative to avoid scale-down requests for consolidation during these periods. However, outside these hours, scale-downs are acceptable.
+
+## Known Requirements and Desired Behaviors 
+**Action and Budget Definition:** Users should be able to define an action and a corresponding budget(s).
+**Supported Actions:** All disruption actions affected by the current Budgets implementation (Consolidation, Emptiness, Expiration, Drift) should be supported.
+**Default Behavior for Unspecified Actions:** Budgets should continue to support a default behavior for all disruption actions. If an action is unspecified, it is assumed to apply to all actions.
+
+### Q: How should Karpenterr handle the default or undefined action case? 
+The current design involves specifying a specific number of disruptable nodes per action, which can complicate the disruption lifecycle. For example, if there's a 10-node budget for "Drift" and a separate 10-node budget for "Consolidation," but a 15-node budget for "All" determining which nodes will get disrupted becomes unclear. Would it be 10 nodes for "Drift" and 5 nodes for "Consolidation"?
+
+#### Unspecified Action as a Global Node Budget
+
+We could consider treating an undefined action as a budget for all disruption actions except for those with explicitly defined budgets. In this scenario, if a user specifies a disruption budget like this:
+
+```yaml
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      action: Consolidation
+      nodes: 10
```

Comment:
I might have missed this but the node for each action's disruption budget is for nodes currently being disrupted by this particular action or all disruption?

Reply:
max that can be disrupted by Consolidation

Reply:
What happens when you go from a higher tolerance period to a lower tolerance period? (This is a problem I just realize is in the existing Node disruption budget design?). Let say during the period of high tolerance we are disrupting nodes at full scale and because disruption currently doesn't have a limit on how long it takes to actually terminate a node, it is possible that many nodes continues to be in the disrupted state when we switch from high to low tolerance. If the expectation of a budget is that this is what a cluster can tolerant during a specific period of time, then we are violating that budget with no recourse? This problem is worsen in complexity here as the "remainder" budget is dynamic which means changes in action budgets becomes harder to track.

Maybe @njtran has some thought.

Reply:
From AKS docs
AKS auto-upgrade didn't upgrade all my agent pools - or one of the pools was upgraded outside of the maintenance window?

If an agent pool fails to upgrade (eg. because of Pod Disruption Budgets preventing it to upgrade) or is in a Failed state, then it might be upgraded later outside of the maintenance window. This scenario is called "catch-up upgrade" and avoids letting Agent pools with a different version than the AKS control plane.


AKS has this scenario where you attempt to upgrade a very large cluster but customer underestimates the window. But they don't want only half of the nodes to be upgraded and would prefer the entire set of nodes gets upgraded. 


Also curious thoughts here, I might create a mega thread for upgrade support and features as there is a lot involved there.

Reply:
Sorry, I know I am pushing back but I like this design and do want this feature. At the end of the day its just an opinion from someone who uses Karpenter so there is no need to gain my approval to move forward.

> The windows are set mainly to say this is when i want to upgrade to start, and customer doesn't care as much when it stops.

I think that kind of behavior is fine but in Kubernetes context + the current naming of "Node Disruption Budget" its problematic. Because most users IMO knows how Pod Disruption Budget works they expect Node Disruption Budget to work the same. That means that a budget defines the max number of disruption that can happen not max number of disruption that can be started. (Note that current implementation follows max disruption not disruption started). This is a very important distinction because as a cluster operator we use the budget to maintain a level of stability for our users.

There is an interesting point about softening or introduction some level of flexibility for starting disruption but I don't think that should be in the Node Disruption Budget

Reply:
> Sorry, I know I am pushing back but I like this design and do want this feature.  
Never apologize for helping someone flesh out a design, the feedback and dicussion is very much so appricated. Thank you for working with me to flesh this out!!

> That means that a budget defines the max number of disruption that can happen not max number of disruption that can be started
It starts to become challenging to block inflight disruption if it starts to exceed the window is my first and main thought. If the vms are already in deleting state  what then?  but it has to be best effort at least and not a promise as there are stages where canceling it would require sending signal to cloud provider to send a cancel request to resource deletion right? Just complicates things quite a bit to stop disruptions already ocurring/triggered.

Reply:
Also to clarify, I believe what you are saying is the right approach where we let all disruptions that took place to continue and during the period of lower tolerance we simply have to wait till the disruption falls back to below the budget. However, I think atm we lack visibility into this and clarify in the docs when describing budgets (i think)